# Session 9

## 7 Other Group of Functions in pandas

### 1- Importing Data



Reading Path 3 ways

In [2]:
import pandas as pd

In [ ]:
path = 'movies.csv'  # csv  comma seperated values
data = pd.read_csv(path)       #mahmoud,25,1990  mahmoud 25 1990
# data

In [ ]:
path = 'D:\Temp/movies.csv'  # Full path csv  comma seperated values
data = pd.read_csv(path)       #mahmoud,25,1990  mahmoud 25 1990
# data

In [3]:
path = '/content/movies.csv'
data = pd.read_csv(path)


In [5]:
data

title     rating  year  runtime
0             Almost Famous          R  2000    122.0
1              American Pie          R  1999     95.0
2        Back to the Future         PG  1985    116.0
3              Blade Runner          R  1982    117.0
4         Blood for Dracula          R  1974    106.0
5               Blue Velvet          R  1986    120.0
6        The Breakfast Club          R  1985     97.0
7                  Clueless      PG-13  1995      NaN
8            Cool Hand Luke         GP  1967    127.0
9                 The Craft          R  1996    101.0
10           Doctor Zhivago      PG-13  1965    197.0
11                  el Topo  Not Rated  1970    125.0
12                Evil Dead      NC-17  1981     85.0
13             Ghostbusters         PG  1984    105.0
14                   Grease      PG-13  1978    110.0
15                 Heathers          R  1988    103.0
16                Labyrinth         PG  1986    101.0
17            The Lost Boys          R  1987     97.0
18               Mean Girls      PG-13  2004     97.0
19       Millennium Actress         PG  2001     87.0
20       My Neighbor Totoro          G  1988     86.0
21        Napoleon Dynamite         PG  2004     96.0
22  Pee-wee's Big Adventure         PG  1985     91.0
23           Pretty in Pink      PG-13  1986     97.0
24       The Princess Bride         PG  1987     98.0
25                   Psycho          R  1960      NaN
26              Stand by Me          R  1986     89.0
27                  Super 8      PG-13  2011    112.0
28                 superbad          R  2007    113.0
29                 WarGames         PG  1983    114.0

In [ ]:
# pd.read_csv(filename)	From a CSV file

In [ ]:
# pd.read_excel(filename)	From an Excel file    xlsx format

In [ ]:
# pd.read_sql(query, connection_object)	Read from a SQL table/database

### 2- Viewing/exploring Data

In [7]:
data

title     rating  year  runtime
0             Almost Famous          R  2000    122.0
1              American Pie          R  1999     95.0
2        Back to the Future         PG  1985    116.0
3              Blade Runner          R  1982    117.0
4         Blood for Dracula          R  1974    106.0
5               Blue Velvet          R  1986    120.0
6        The Breakfast Club          R  1985     97.0
7                  Clueless      PG-13  1995      NaN
8            Cool Hand Luke         GP  1967    127.0
9                 The Craft          R  1996    101.0
10           Doctor Zhivago      PG-13  1965    197.0
11                  el Topo  Not Rated  1970    125.0
12                Evil Dead      NC-17  1981     85.0
13             Ghostbusters         PG  1984    105.0
14                   Grease      PG-13  1978    110.0
15                 Heathers          R  1988    103.0
16                Labyrinth         PG  1986    101.0
17            The Lost Boys          R  1987     97.0
18               Mean Girls      PG-13  2004     97.0
19       Millennium Actress         PG  2001     87.0
20       My Neighbor Totoro          G  1988     86.0
21        Napoleon Dynamite         PG  2004     96.0
22  Pee-wee's Big Adventure         PG  1985     91.0
23           Pretty in Pink      PG-13  1986     97.0
24       The Princess Bride         PG  1987     98.0
25                   Psycho          R  1960      NaN
26              Stand by Me          R  1986     89.0
27                  Super 8      PG-13  2011    112.0
28                 superbad          R  2007    113.0
29                 WarGames         PG  1983    114.0

In [ ]:
df = data

In [8]:
df = data.copy()   # copy because not to overwrite orignial data

 Solution

create a deep copy of the DataFrame. This ensures that any modifications made to the new DataFrame (df) do not affect the original DataFrame (data).

Here’s a detailed explanation:

- Avoid Unintended Modifications: When you assign df = data, both df and data refer to the same DataFrame in memory. Any changes made to df will also affect data and vice versa. By using df = data.copy(), you create a separate copy of the DataFrame. This allows you to modify df without altering data.

- Preserve Original Data: Keeping the original data unchanged is crucial, especially when you need to compare the modified DataFrame with the original or when the original DataFrame is required for further operations.

- Ensure Data Integrity: In data analysis and preprocessing, it’s common to perform various transformations and experiments on the data. Creating a copy helps ensure that you can always revert to the original dataset if needed.

In [ ]:
# df = data     # Forbidden

In [9]:
df.head() #First n rows of the DataFrame

title rating  year  runtime
0       Almost Famous      R  2000    122.0
1        American Pie      R  1999     95.0
2  Back to the Future     PG  1985    116.0
3        Blade Runner      R  1982    117.0
4   Blood for Dracula      R  1974    106.0

In [ ]:
df.tail() #Last n rows of the DataFrame

title rating  year  runtime
25       Psycho      R  1960      NaN
26  Stand by Me      R  1986     89.0
27      Super 8  PG-13  2011    112.0
28     superbad      R  2007    113.0
29     WarGames     PG  1983    114.0

In [12]:
df.sample(5)

title rating  year  runtime
22  Pee-wee's Big Adventure     PG  1985     91.0
24       The Princess Bride     PG  1987     98.0
18               Mean Girls  PG-13  2004     97.0
25                   Psycho      R  1960      NaN
13             Ghostbusters     PG  1984    105.0

In [13]:
df.shape   #Number of rows and columns

(30, 4)

In [ ]:
df.info() #Index, Datatype and Memory information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   title    30 non-null     object 
 1   rating   30 non-null     object 
 2   year     30 non-null     int64  
 3   runtime  28 non-null     float64
dtypes: float64(1), int64(1), object(2)
memory usage: 1.1+ KB


**Insights**

- 1- data has 30 rows and 4 columns

- 2- range index (0 to 29) means there is no missing data

- 3- column runtime has 2 nulls

- 4- data has 2 object data type

In [ ]:
df['rating'].value_counts() # apply only on categorical (string) features

rating
R            12
PG            8
PG-13         6
GP            1
Not Rated     1
NC-17         1
G             1
Name: count, dtype: int64

Task

how to get percentage of counting for every value

In [15]:
df['rating'].value_counts() / len(df)

rating
R            0.400000
PG           0.266667
PG-13        0.200000
GP           0.033333
Not Rated    0.033333
NC-17        0.033333
G            0.033333
Name: count, dtype: float64

In [ ]:
df['rating'].value_counts(normalize =True)

rating
R            0.400000
PG           0.266667
PG-13        0.200000
GP           0.033333
Not Rated    0.033333
NC-17        0.033333
G            0.033333
Name: proportion, dtype: float64

R rating is the most repeated

task

1- read dataframe

2- display information

3- print last 5 rows

4- print shape of it

5- write info of data

6- print count for any unique values

### 3- Statistics

In [16]:
df.describe(include='all')   #Summary statistics for all columns

title rating         year     runtime
count              30     30    30.000000   28.000000
unique             30      7          NaN         NaN
top     Almost Famous      R          NaN         NaN
freq                1     12          NaN         NaN
mean              NaN    NaN  1987.000000  107.285714
std               NaN    NaN    12.406895   21.308766
min               NaN    NaN  1960.000000   85.000000
25%               NaN    NaN  1982.250000   96.750000
50%               NaN    NaN  1986.000000  102.000000
75%               NaN    NaN  1995.750000  114.500000
max               NaN    NaN  2011.000000  197.000000

**Insights**

- 1- title feature has 30 values with 30 unique values
- 2- Rating Feature has 30 values with 7 unique values
- 3- Rating Feature has the most freq value is R with 12 times
- 4- min and max of year feature is 1960 and 2011 respectively ......
- 5- runtime feature has 28 values with 2 missing nulls.
- 6- the average of time of movies is almost 107 mintues
- 7- the shortest movie is 85 minute with it is name  .....
- 8- the longest movie with 197 mintue is .....

Task

create 3 insights (statistics) from Ecommerce Super market

Task  :

1- find the shortest and longest movie name

2- which movie is the oldest and newest

In [ ]:
df['runtime'].min()

85.0

In [17]:
df.loc[ df['runtime'] == df['runtime'].min() ,'title']

title rating  year  runtime
12  Evil Dead  NC-17  1981     85.0

In [ ]:
df[ df['runtime'] == df['runtime'].min() ]['title']

12    Evil Dead
Name: title, dtype: object

In [ ]:
df[ df['runtime'] == df['runtime'].min() ]['title'].values[0]

'Evil Dead'

In [ ]:
# df['title']

In [ ]:
df[df['runtime'] == df['runtime'].min() ]['title']

12    Evil Dead
Name: title, dtype: object

In [18]:
df[df['runtime'] == df['runtime'].max() ]

title rating  year  runtime
10  Doctor Zhivago  PG-13  1965    197.0

In [19]:
df[['title','runtime']]

title  runtime
0             Almost Famous    122.0
1              American Pie     95.0
2        Back to the Future    116.0
3              Blade Runner    117.0
4         Blood for Dracula    106.0
5               Blue Velvet    120.0
6        The Breakfast Club     97.0
7                  Clueless      NaN
8            Cool Hand Luke    127.0
9                 The Craft    101.0
10           Doctor Zhivago    197.0
11                  el Topo    125.0
12                Evil Dead     85.0
13             Ghostbusters    105.0
14                   Grease    110.0
15                 Heathers    103.0
16                Labyrinth    101.0
17            The Lost Boys     97.0
18               Mean Girls     97.0
19       Millennium Actress     87.0
20       My Neighbor Totoro     86.0
21        Napoleon Dynamite     96.0
22  Pee-wee's Big Adventure     91.0
23           Pretty in Pink     97.0
24       The Princess Bride     98.0
25                   Psycho      NaN
26              Stand by Me     89.0
27                  Super 8    112.0
28                 superbad    113.0
29                 WarGames    114.0

In [ ]:
df[df['runtime'] == df['runtime'].max() ][['title','runtime']]

title  runtime
10  Doctor Zhivago    197.0

In [ ]:
df[df['year'] == df['year'].min() ][['title','year']]

title  year
25  Psycho  1960

In [ ]:
df[df['year'] == df['year'].max() ][['title','year']]

title  year
27  Super 8  2011

In [ ]:
df['runtime'].mean()	#Returns the mean of all columns

107.28571428571429

In [ ]:
df['runtime'].count()	#Returns the number of non-null values in each DataFrame column

28

In [ ]:
df['runtime'].max()	#Returns the highest value in each column

197.0

In [ ]:
df['runtime'].min()	#Returns the lowest value in each column

85.0

In [ ]:
df['runtime'].median()	#Returns the median of each column

102.0

In [ ]:
df['runtime'].std()	#Returns the standard deviation of each column

21.308765615668406

In [20]:
df

title     rating  year  runtime
0             Almost Famous          R  2000    122.0
1              American Pie          R  1999     95.0
2        Back to the Future         PG  1985    116.0
3              Blade Runner          R  1982    117.0
4         Blood for Dracula          R  1974    106.0
5               Blue Velvet          R  1986    120.0
6        The Breakfast Club          R  1985     97.0
7                  Clueless      PG-13  1995      NaN
8            Cool Hand Luke         GP  1967    127.0
9                 The Craft          R  1996    101.0
10           Doctor Zhivago      PG-13  1965    197.0
11                  el Topo  Not Rated  1970    125.0
12                Evil Dead      NC-17  1981     85.0
13             Ghostbusters         PG  1984    105.0
14                   Grease      PG-13  1978    110.0
15                 Heathers          R  1988    103.0
16                Labyrinth         PG  1986    101.0
17            The Lost Boys          R  1987     97.0
18               Mean Girls      PG-13  2004     97.0
19       Millennium Actress         PG  2001     87.0
20       My Neighbor Totoro          G  1988     86.0
21        Napoleon Dynamite         PG  2004     96.0
22  Pee-wee's Big Adventure         PG  1985     91.0
23           Pretty in Pink      PG-13  1986     97.0
24       The Princess Bride         PG  1987     98.0
25                   Psycho          R  1960      NaN
26              Stand by Me          R  1986     89.0
27                  Super 8      PG-13  2011    112.0
28                 superbad          R  2007    113.0
29                 WarGames         PG  1983    114.0

task Data anaylsis


- average of  movies with rating R

- calcualte the average of movies period that published before 1987

Bonus

- get movies  that runtime <90 or rating = R

- calcualte the max of movies period that published after 1995 and rating R

In [ ]:
df[df['rating'] == 'R']['runtime'].mean()

105.45454545454545

In [ ]:
df[df['year'] < 1987]['runtime'].mean()

112.3125